In [33]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [34]:
car_df=pd.read_csv(r"C:\Users\RBI\Downloads\archive (4)\car data.csv")

In [35]:
car_df

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0       ritz  2014           3.35           5.59       27000    Petrol   
1        sx4  2013           4.75           9.54       43000    Diesel   
2       ciaz  2017           7.25           9.85        6900    Petrol   
3    wagon r  2011           2.85           4.15        5200    Petrol   
4      swift  2014           4.60           6.87       42450    Diesel   
..       ...   ...            ...            ...         ...       ...   
296     city  2016           9.50          11.60       33988    Diesel   
297     brio  2015           4.00           5.90       60000    Petrol   
298     city  2009           3.35          11.00       87934    Petrol   
299     city  2017          11.50          12.50        9000    Diesel   
300     brio  2016           5.30           5.90        5464    Petrol   

    Selling_type Transmission  Owner  
0         Dealer       Manual      0  
1         Dealer       Manual      0  
2         Dealer       Manual      0  
3         Dealer       Manual      0  
4         Dealer       Manual      0  
..           ...          ...    ...  
296       Dealer       Manual      0  
297       Dealer       Manual      0  
298       Dealer       Manual      0  
299       Dealer       Manual      0  
300       Dealer       Manual      0  

[301 rows x 9 columns]

### Preprocessing the data

In [36]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Driven_kms     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Selling_type   301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [37]:
car_df.isnull().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Owner            0
dtype: int64

checking for nul values. There are no null values present.

In [38]:
car_df.duplicated().sum()


2

There are two duplicated values and we have to drop them.

In [39]:
car_df.drop_duplicates(inplace=True)

In [40]:
car_df.shape

(299, 9)

In [41]:
car_df.describe()

Year  Selling_Price  Present_Price     Driven_kms       Owner
count   299.000000     299.000000     299.000000     299.000000  299.000000
mean   2013.615385       4.589632       7.541037   36916.752508    0.043478
std       2.896868       4.984240       8.566332   39015.170352    0.248720
min    2003.000000       0.100000       0.320000     500.000000    0.000000
25%    2012.000000       0.850000       1.200000   15000.000000    0.000000
50%    2014.000000       3.510000       6.100000   32000.000000    0.000000
75%    2016.000000       6.000000       9.840000   48883.500000    0.000000
max    2018.000000      35.000000      92.600000  500000.000000    3.000000

In [42]:
# Converting categorical data into numerical data

colname=[]
for x in car_df.columns:
    if car_df[x].dtype=='object':
        colname.append(x)
colname

['Car_Name', 'Fuel_Type', 'Selling_type', 'Transmission']

In [43]:
from sklearn.preprocessing import LabelEncoder
 
le=LabelEncoder()
 
for x in colname:
    car_df[x]=le.fit_transform(car_df[x])

In [44]:
car_df.head()

Car_Name  Year  Selling_Price  Present_Price  Driven_kms  Fuel_Type  \
0        90  2014           3.35           5.59       27000          2   
1        93  2013           4.75           9.54       43000          1   
2        68  2017           7.25           9.85        6900          2   
3        96  2011           2.85           4.15        5200          2   
4        92  2014           4.60           6.87       42450          1   

   Selling_type  Transmission  Owner  
0             0             1      0  
1             0             1      0  
2             0             1      0  
3             0             1      0  
4             0             1      0

In [45]:
# creating X & Y

X= car_df.loc[:,['Car_Name', 'Year', 'Present_Price', 'Driven_kms','Fuel_Type', 'Selling_type', 'Transmission', 'Owner']]
Y= car_df.loc[:,['Selling_Price']]

In [46]:
# Scaling the data

from sklearn.preprocessing import StandardScaler

scaler= StandardScaler()

scaler.fit(X)
X= scaler.transform(X)

In [47]:
#Split the data into test and train
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,      # 70:30 is split ratio
                                                    random_state=10)  

### Random Forest 

In [48]:
# Predicting using the RandomForesrtRegressor

from sklearn.ensemble import RandomForestRegressor

model_RF=RandomForestRegressor(n_estimators=500,criterion="mse",random_state=10)
model_RF.fit(X_train,Y_train)

Y_pred=model_DecisionTree.predict(X_test)
print(Y_pred)

[ 2.8059   4.266    1.17226  0.41442  4.6628   0.62328  0.24658  0.52446
  4.5619   7.12742  4.79752  2.22266  0.3509   9.15862  9.05814  4.9838
 16.78348  2.8968   0.68756  8.77976 15.27594  0.23694  8.351    0.99018
  1.2042   0.98568  0.3883   4.4548   1.01486  0.7627   5.829    0.31742
  3.75252  0.4754   0.27682  1.119    4.02402  5.40278  0.75824  0.3597
  3.50702  5.99102  5.17882  9.8269   9.80148  0.3371   0.55674  1.05058
  1.06344  2.4983   1.4131   2.61212  0.46278  4.0818   0.32924  2.9576
  0.43816  7.33348  6.338    4.5086 ]


In [49]:
from sklearn.metrics import r2_score,mean_squared_error
import numpy as np

r2=r2_score(Y_test,Y_pred)
print("R-squared:",r2)

rmse=np.sqrt(mean_squared_error(Y_test,Y_pred))
print("RMSE:",rmse)

adjusted_r_squared = 1 - (1-r2)*(len(Y)-1)/(len(Y)-X.shape[1]-1)
print("Adj R-square:",adjusted_r_squared)

R-squared: 0.965074602764954
RMSE: 0.7436114348591855
Adj R-square: 0.9641111435308838


In [50]:
# score value
model_RF.score(X_train, Y_train)

0.9870783316846026

# Conclusion

The Random Forest model gives a good prediction.

The R-squared and Adj R-squared values are 0.96507 & 0.96411 simultaneously. 

It is also close to score value.